### PR TITLE
fix bug with compiledProgram

### PR DESCRIPTION
**Pull request by seiriosPlus**
add thread barrier in ssa_graph_executor, for sync/half_async with the communicator.

### CODE REVIEW

**Review by MrChengmo — APPROVED** [review]
LGTM

File: paddle/fluid/pybind/pybind.cc [hunk 1]
```diff
@@ -1732,6 +1732,14 @@ All parameter, weight, gradient are variables in Paddle.
           R"DOC(This config that how many iteration the executor will run when
                 user call exe.run() in python
               )DOC")
+      .def_property(
+          "use_thread_barrier",
+          [](const ExecutionStrategy &self) { return self.thread_barrier_; },
+          [](ExecutionStrategy &self, bool use_thread_barrier) {
+            self.thread_barrier_ = use_thread_barrier;
+          },
+          R"DOC(This config that the this is distributed training with parameter server
```
Comment (guru4elephant) [inline paddle/fluid/pybind/pybind.cc:1741]: I think this doc can be more detailed.

File: python/paddle/fluid/incubate/fleet/parameter_server/distribute_transpiler/distributed_strategy.py [hunk 2]
```diff
@@ -196,8 +196,9 @@ def __init__(self):
         super(HalfAsyncStrategy, self).__init__()
         self._program_config.sync_mode = False
         self._program_config.runtime_split_send_recv = True
-        self._build_strategy.async_mode = True
         self._program_config.half_async = True
+        self._build_strategy.async_mode = True
+        self._execute_strategy.use_thread_barrier = True
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/parameter_server/distribute_transpiler/distributed_strategy.py:201]: what if async_mode is False here?

Comment (seiriosPlus) [reply python/paddle/fluid/incubate/fleet/parameter_server/distribute_transpiler/distributed_strategy.py:201]: AsyncMode=False 表示使用SyncSSAGraph， 线程间的同步控制是由SSA GRAPH来控制的，用Reduce/AllReduce来进行阻塞参数的同步。 

AsyncMode=True 表示使用AsyncSSAGraph，线程间不再等待， 线程的同步则交给Communicator的Barreir来控制。

File: paddle/fluid/pybind/pybind.cc (hunk 1 above)
Comment (seiriosPlus) [reply paddle/fluid/pybind/pybind.cc:1741]: CI is very difficult, I will modify the doc in the next PR.

**Pull request merged**